Signal/src/ViewControllers/MessageComposeTableViewController.m

```diff
@@ -168,275 +194,229 @@ - (void)viewWillAppear:(BOOL)animated
-                                    forState:UIControlStateNormal];
-    [inviteButton.titleLabel setFont:[UIFont ows_regularFontWithSize:17.f]];
+#pragma mark - Table Contents
 
-    UIButton *searchByPhoneNumberButton = [UIButton buttonWithType:UIButtonTypeCustom];
-    [searchByPhoneNumberButton setTitle:NSLocalizedString(@"NO_CONTACTS_SEARCH_BY_PHONE_NUMBER",
-                                                          @"Label for a button that lets users search for contacts by phone number")
-                               forState:UIControlStateNormal];
-    [searchByPhoneNumberButton setTitleColor:[UIColor ows_materialBlueColor]
-                                    forState:UIControlStateNormal];
-    [searchByPhoneNumberButton.titleLabel setFont:[UIFont ows_regularFontWithSize:17.f]];
-    [inviteButton.superview addSubview:searchByPhoneNumberButton];
-    [searchByPhoneNumberButton autoHCenterInSuperview];
-    [searchByPhoneNumberButton autoPinEdge:ALEdgeTop
-                                    toEdge:ALEdgeBottom
-                                    ofView:inviteButton
-                                withOffset:20];
-    [searchByPhoneNumberButton addTarget:self
-                                  action:@selector(hideBackgroundView)
-                        forControlEvents:UIControlEventTouchUpInside];
-}
-
-- (void)hideBackgroundView {
-    [[Environment preferences] setHasDeclinedNoContactsView:YES];
-    
-    [self showEmptyBackgroundViewIfNecessary];
-}
-
-- (void)presentInviteFlow
+- (void)updateTableContents
 {
-    OWSInviteFlow *inviteFlow =
-        [[OWSInviteFlow alloc] initWithPresentingViewController:self contactsManager:self.contactsManager];
-    [self presentViewController:inviteFlow.actionSheetController animated:YES completion:nil];
-}
+    OWSTableContents *contents = [OWSTableContents new];
 
-- (void)showLoadingBackgroundView:(BOOL)show {
-    if (show) {
-        self.searchController.searchBar.hidden = YES;
-        self.tableView.backgroundView          = _loadingBackgroundView;
-        self.tableView.backgroundView.opaque   = YES;
-    } else {
-        self.searchController.searchBar.hidden = NO;
-        self.tableView.backgroundView          = nil;
+    if (self.isNoContactsModeActive) {
+        self.tableViewController.contents = contents;
+        return;
     }
-}
-
-- (void)showEmptyBackgroundViewIfNecessary {
-    self.isNoContactsViewVisible = ([self.contacts count] == 0 &&
-                                    ![[Environment preferences] hasDeclinedNoContactsView]);
-}
 
-- (void)setIsNoContactsViewVisible:(BOOL)isNoContactsViewVisible {
-    if (isNoContactsViewVisible == _isNoContactsViewVisible) {
-        return;
+    __weak MessageComposeTableViewController *weakSelf = self;
+    ContactsViewHelper *helper = self.contactsViewHelper;
+
+    OWSTableSection *section = [OWSTableSection new];
+
+    // Find Non-Contacts by Phone Number
+    [section addItem:[OWSTableItem itemWithCustomCellBlock:^{
```

Comment:
As you pointed out, the main reason I included "find by phone number" is that it provides a much better experience for finding people in other calling codes.  Very few people are comfortable with e164 formatting / dealing with country codes.  The "find by phone number" view has controls that let surface this, let you specify the calling code by country name, etc. - and it also takes care of the concatenation process.  

I also like that its presence emphasizes that you can communicate with non-contacts - the gaze tends to gloss over placeholder text.

I do think we should leave the ability to search for phone numbers in the search bar.

I'm not worried about the redundancy - offering more than one way to do things seems fine to me.  

> Also, it's taking up some real prime real estate, especially on smaller devices:

Yes, that's an issue.

> Another thing, when I type in part of a phone number, e.g. "555-123-12", "Find by phone number" almost seems like a button to execute the search for accounts based on my incomplete search input.

That's true, but if anyone made that mistake I think they'd figure it out pretty quickly, close the "find by phone number" view and not have lost their search.

> Weighing those things, I'm not sure we should keep it. WDYT @charlesmchen?

It's not perfect, but I think it's the lesser of two evils to keep it in.  The "affordance" of being able to search for non-contacts by phone number isn't clear enough without it and finding users in other calling codes is currently a bad experience.

Let's discuss on a call!

> In either case, on a separate design note, I believe the top two items probably are the same height as the rest of the table cells, but without the avatars, the top two items feel a little too big. Do you think it would be unreasonable to make them shorter than the other items in the table?

I had it that way originally and it felt broken to have them the default size.  Maybe there's a happy medium in between the two cell heights.